### ISSUE TITLE
useMethods

### ISSUE THREAD

**Issue by atomiks** [streamich/react-use#206]
Similar to: https://github.com/pelotom/use-methods

Neat alternative to `useReducer`

## Hook

```tsx
function useMethods(createMethods, initialState) {
  const reducer = useMemo(
    () => (state, action) => {
      return createMethods(state)[action.type](...action.payload);
    },
    [createMethods]
  );
  const [state, dispatch] = useReducer(reducer, initialState);
  const actionTypes = Object.keys(createMethods(initialState));
  const wrappedMethods = useMemo(() => {
    return actionTypes.reduce((acc, type) => {
      acc[type] = (...payload) => dispatch({ type, payload });
      return acc;
    }, {});
  }, actionTypes);

  return [state, wrappedMethods];
}
```

## Usage

```tsx
const initialState = {
  count: 0,
};

function createMethods(state) {
  return {
    reset() {
      return initialState;
    },
    increment() {
      return { ...state, count: state.count + 1 };
    },
    decrement() {
      return { ...state, count: state.count - 1 };
    },
  };
}

function Counter() {
  const [state, methods] = useMethods(createMethods, initialState);

  return (
    <>
      Count: {state.count}
      <button onClick={methods.increment}>+</button>
      <button onClick={methods.decrement}>-</button>
    </>
  );
}
```

**Comment by wardoost** [comment]
This looks pretty neat indeed. Is there also a way to use async methods with this?

**Comment by atomiks** [comment]
> Is there also a way to use async methods with this?

Perhaps return a function that receives the methods object like this:

```jsx
const createMethods = state => ({
  incrementAsync: () => async methods => {
    await new Promise(done => setTimeout(done, 1000));
    methods.increment();
  }),
  increment: () => ({ ...state, count: state.count + 1 })
});
```

So you will call the sync reducers from the async methods.

```jsx

function useMethods(createMethods, initialState) {
  const reducer = useMemo(
    () => (state, action) => {
      return createMethods(state)[action.type](...action.payload);
    },
    [createMethods]
  );
  const [state, dispatch] = useReducer(reducer, initialState);
  const methods = createMethods(initialState);
  const actionTypes = Object.keys(methods);
  const wrappedMethods = useMemo(() => {
    return actionTypes.reduce((acc, type) => {
      acc[type] =
        typeof methods[type]() === 'function'
          ? (...payload) => methods[type](...payload)(acc)
          : (...payload) => dispatch({ type, payload });
      return acc;
    }, {});
  }, [actionTypes]);

  return [state, wrappedMethods];
}
```

**Comment by wardoost** [comment]
Yeah something like that, and maybe add a second getState function parameter so it's more similar to [`redux-thunk`](https://github.com/reduxjs/redux-thunk). We could actually generalise the hook to take any redux middleware as suggested in #164

**Comment by mathieutu** [comment]
Hi folks, 
I'm seeing this issue, and that the useMethod hook is present in the package, but undocumented. What is the state of that? Is it ready from prod usage?

Thanks.

**Comment by ayush987goyal** [comment]
@streamich @atomiks  Hey! Could you please review the above PR?

**Comment by ayush987goyal** [comment]
@streamich Merged with #910
